contrib/pam_zfs_key/pam_zfs_key.c

```diff
@@ -0,0 +1,600 @@
+#include <sys/zfs_context.h>
+#include <sys/fs/zfs.h>
+#include <sys/dsl_crypt.h>
+#include <sys/byteorder.h>
+#include <libzfs.h>
+
+#include <syslog.h>
+
+#include <sys/zio_crypt.h>
+#include <openssl/evp.h>
+
+#define	PAM_SM_AUTH
+#define	PAM_SM_PASSWORD
+#define	PAM_SM_SESSION
+#include <security/pam_modules.h>
+#include <security/pam_ext.h>
+#include <string.h>
+#include <signal.h>
+
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/file.h>
+#include <sys/wait.h>
+#include <pwd.h>
+
+#include <sys/mman.h>
+static void destroy_pw(pam_handle_t *pamh, void *data, int errcode);
+typedef struct {
+	size_t len;
+	char *value;
+} password;
+
+static password *alloc_pw_size(size_t len) {
+	password *pw = malloc(sizeof (password));
+	pw->len = len;
+	pw->value = malloc(len);
+	mlock(pw->value, pw->len);
+	return (pw);
+}
```

Comment:
style nit: for consistency with the function declaration in the rest of the project, please place both the return type and opening `{` on their own lines, so so.   Also make sure to include a new line between functions.

```c
static password *
alloc_pw_size(size_t len)
{
	password *pw = malloc(sizeof (password));
	pw->len = len;
	pw->value = malloc(len);
	mlock(pw->value, pw->len);
	return (pw);
}

static password *
lloc_pw_string(const char *source) {
...
}
```